src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3500,651 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
   return query;
 }
 
+/*
+ * qo_reduce_joined_referenced_tables () - Removes the table of the primary key from the join
+ *                                         with the table with the foreign key referencing it.
+ *   return: void
+ *   parser(in): parser context
+ *   query(in): query to check
+ *
+ * Note: The table with the primary key can be removed from the join with the table with the foreign key referencing it
+ *       if there is no other reference than the joining purpose. This is because the relationship
+ *       between the primary key and the foreign key replaces the data filter of the join. If the join of the table
+ *       with the primary key is removed, predicates for IS NOT NULL must be added because it cannot be filtered
+ *       for NULL.
+ *
+ *    e.g. drop if exists child, parent;
+ *         create table parent (c1 int primary key, c2 int);
+ *         create table child (c1 int, parent_c1 int references parent (c1), c2 int);
+ *
+ *         select c.* from child c inner join parent p on p.c1 = c.parent_c1 where c.c2 = 1;
+ *         select c.* from child c, parent p where c.parent_c1 = p.c1 and c.c2 = 1;
+ *           -> select c.* from child c where c.c2 = 1 and c.parent_c1 is not null;
+ *
+ *         select c.* from child c inner join parent p on p.c1 = c.parent_c1 where p.c2 = 1;
+ *         select c.* from child c, parent p where c.parent_c1 = p.c1 and p.c2 = 1;
+ *           -> do not change.
+ */
+static void
+qo_reduce_joined_referenced_tables (PARSER_CONTEXT * parser, PT_NODE * query)
```

Comment:
Could you divide this function? This function has a long LOC, and the while-loops and for-loops are nested. So It is hard to follow the routine.

Reply:
I have separated the function as below.
```
qo_reduce_joined_tables_referenced_by_foreign_key()
  -> qo_is_exclude_spec()
  -> qo_check_primary_key_referenced_by_foreign_key_in_parent_spec()
  -> qo_check_foreign_keys_referencing_primary_key_in_child_spec()
    -> qo_check_foreign_key_referencing_primary_key_in_child_spec()
  -> qo_reduce_predicate_for_parent_spec()
```